### PR TITLE
add JDBC DAO implementations

### PR DESCRIPTION
**Pull request by Alkey**

### CODE REVIEW

**Review by KostiantynVorobiov — APPROVED** [review]
it's ok

File: src/main/java/com/internet/shop/dao/impl/OrderJdbcDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    @Override
+    public Order update(Order order) {
+        deleteProductsFromOrder(order.getId());
+        addProductsToOrder(order);
+        return order;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        deleteProductsFromOrder(id);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/OrderJdbcDaoImpl.java]: don't delete products, mark order deleted, that would be enough

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                shoppingCart.setId(resultSet.getLong("cart_id"));
+                shoppingCart.setUserId(resultSet.getLong("user_id"));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't find ShoppingCart by user ID: " + id, e);
+        }
+        getShoppingCartProducts(shoppingCart);
+        return Optional.of(shoppingCart);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java:36]: in this method you will never return optional empty, but we should do that if cart was not found

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        try (PreparedStatement statement = ConnectionUtil.getConnection()
+                .prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {
+            statement.setLong(1, shoppingCart.getUserId());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                shoppingCart.setId(resultSet.getLong(1));
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java]: e.printStackTrace() is a bad practice

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                shoppingCart.setId(resultSet.getLong("cart_id"));
+                shoppingCart.setUserId(resultSet.getLong("user_id"));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't find shopping cart by ID: " + id, e);
+        }
+        getShoppingCartProducts(shoppingCart);
+        return Optional.of(shoppingCart);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java:74]: same here with optional

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    @Override
+    public ShoppingCart update(ShoppingCart shoppingCart) {
+        deleteProductsFromShoppingCart(shoppingCart.getId());
+        addProductsToShoppingCart(shoppingCart);
+        return shoppingCart;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        deleteProductsFromShoppingCart(id);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java]: no need to delete products, just mark cart as deleted

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        List<ShoppingCart> shoppingCarts = new ArrayList<>();
+        try (PreparedStatement statement = ConnectionUtil.getConnection().prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                ShoppingCart shoppingCart = new ShoppingCart();
+                shoppingCart.setId(resultSet.getLong("cart_id"));
+                shoppingCart.setUserId(resultSet.getLong("user_id"));
+                shoppingCarts.add(shoppingCart);
+            }
+            shoppingCarts.forEach(this::getShoppingCartProducts);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java]: connection inside of connection

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            statement.setLong(1, shoppingCart.getId());
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                products.add(new Product(resultSet.getLong("product_id"),
+                        resultSet.getString("name"),
+                        resultSet.getDouble("price")));
+            }
+            shoppingCart.setProducts(products);
+        } catch (SQLException e) {
+            e.printStackTrace();
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/ShoppingCartJdbcDaoImpl.java]: same with printStackTrace()

File: src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            return Optional.ofNullable(createUser(statement.executeQuery()));
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get user from DB by ID: " + id, e);
+        }
+    }
+
+    @Override
+    public User update(User user) {
+        deleteUserRoles(user.getId());
+        String query = "UPDATE users SET name = ?, login = ?, password = ? WHERE user_id = ?";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java]: we don't want to update deleted user

File: src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+        try (PreparedStatement statement = ConnectionUtil.getConnection().prepareStatement(query)) {
+            statement.setString(1, user.getName());
+            statement.setString(2, user.getLogin());
+            statement.setString(3, user.getPassword());
+            statement.setLong(4, user.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't update user", e);
+        }
+        setRoleIds(user.getRoles());
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java]: this line is redundant, roles without id can be only in create method

File: src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        addUserRoles(user);
+        return user;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        deleteUserRoles(id);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java]: don't delete roles, just mark user as deleted

File: src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't add user roles to DB", e);
+        }
+    }
+
+    private void setRoleIds(Set<Role> userRoles) {
+        String query = "SELECT * FROM roles";
+        try (PreparedStatement statement = ConnectionUtil.getConnection().prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java:143]: you can do that with one loop, change query to select * from roles where role_name = ? 
you will need only one loop then

File: src/main/java/com/internet/shop/dao/impl/OrderJdbcDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+@Dao
+public class OrderJdbcDaoImpl implements OrderDao {
+    @Override
+    public List<Order> getUserOrders(Long userId) {
+        String query = "SELECT * FROM orders WHERE user_id = ? AND deleted = false";
+        List<Order> ordersFromDB;
+        try (PreparedStatement statement = ConnectionUtil.getConnection()
+                .prepareStatement(query)) {
+            statement.setLong(1, userId);
+            ordersFromDB = getOrdersFromDB(statement.executeQuery());
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/impl/OrderJdbcDaoImpl.java:26]:
```suggestion
            ordersFromDB = getOrdersFromResultSet(statement.executeQuery());
```

File: src/main/java/com/internet/shop/dao/impl/OrderJdbcDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+            throw new DataProcessingException("Can't create order", e);
+        }
+        addProductsToOrder(order);
+        return order;
+    }
+
+    @Override
+    public Optional<Order> get(Long id) {
+        String query = "SELECT * FROM orders WHERE order_id = ? AND deleted = false";
+        Order order = new Order();
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/impl/OrderJdbcDaoImpl.java:55]: in this case you will alwasy return the Order even when it's has not been found in DB. Let's create it only if resultSet.next() == true.

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**